ssl/ssl_key_share.cc

```diff
@@ -381,6 +383,180 @@ class CECPQ2bKeyShare : public SSLKeyShare {
   uint8_t public_sike_[SIKE_PUB_BYTESZ];
 };
 
+// KeyShare class for OQS supplied hybrid and post-quantum crypto algs
+// Hybrid messages are encoded as follows: classical_len | classical_artifact | pq_len | pq_artifact
+// TODOs (FIXMEOQS):
+// * make sure the format is the same as in the OpenSSL fork, i.e., put length of key, etc.
+class OQSKeyShare : public SSLKeyShare {
+ public:
+  // Although oqs_meth can be determined from the group_id,
+  // we pass both in as the translation from group_id to
+  // oqs_meth is already done by SLKeyShare::Create
+  // to determine whether oqs_meth is enabled in liboqs
+  // and return nullptr if not. It is easier to handle
+  // the error in there as opposed to in this constructor.
+  OQSKeyShare(uint16_t group_id, const char *oqs_meth, bool is_hybrid) : group_id_(group_id), is_hybrid_(is_hybrid) {
+    pq_kex_= OQS_KEM_new(oqs_meth);
+    if (is_hybrid_) {
+      classical_kex_ = SSLKeyShare::Create(SSL_CURVE_SECP256R1);
+    }
+  }
+
+  uint16_t GroupID() const override { return group_id_; }
+
+  bool Offer(CBB *out) override {
+    Array<uint8_t> classical_public_key;
+    Array<uint8_t> pq_public_key;
+
+    // For a hybrid KEX, generate the classical keys first
+    if (is_hybrid_) {
+      ScopedCBB classical_offer;
+      if (!CBB_init(classical_offer.get(), p256_public_key_size_) ||
+          !classical_kex_->Offer(classical_offer.get()) ||
+          !CBBFinishArray(classical_offer.get(), &classical_public_key)) {
+        // the classical code will set the appropriate error on failure
+        return false;
+      }
+    }
+    // Generate the PQ key pair.
+    if (!pq_public_key.Init(pq_kex_->length_public_key) ||
```

Comment:
Do we have to worry about deallocated `pq_public_key` or `pq_private_key` later, or is that handled automatically?

Reply:
`pq_public_key` is an `Array` object (which is defined by BoringSSL). As per the "Storage reuse" section [here](https://en.cppreference.com/w/cpp/language/lifetime), we don't have to explicitly call the object destructor since we do not rely on its side effects. For all the `CBB`s we define we call `CBBFinishArray`, which should be enough.